sudachi/src/analysis/morpheme_list.rs

```diff
@@ -0,0 +1,275 @@
+/*
+ *  Copyright (c) 2021 Works Applications Co., Ltd.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+use std::ops::Deref;
+
+use crate::analysis::node::Node;
+use crate::analysis::stateless_tokenizer::DictionaryAccess;
+use crate::dic::grammar::Grammar;
+use crate::dic::lexicon::word_infos::WordInfo;
+use crate::input_text::Utf8InputText;
+use crate::prelude::*;
+
+pub struct MorphemeList<T> {
+    pub dict: T,
+    pub input_text: String,
+    pub path: Vec<Node>,
+}
+
+impl<T> MorphemeList<T>
+where
+    T: Deref,
+    <T as Deref>::Target: DictionaryAccess,
+{
+    pub fn new(dict: T, input_text: &Utf8InputText, mut path: Vec<Node>) -> SudachiResult<Self> {
+        for node in &mut path {
+            node.fill_word_info(dict.lexicon())?;
+            // set range for origina text
+            node.set_range(
+                input_text.get_original_index(node.begin),
+                input_text.get_original_index(node.end),
+            )
+        }
+
+        Ok(Self {
+            dict,
+            input_text: input_text.original.to_string(),
+            path,
+        })
+    }
+
+    pub fn empty(dict: T) -> Self {
+        Self {
+            dict,
+            input_text: String::new(),
+            path: Vec::new(),
+        }
+    }
+
+    fn get_grammar(&self) -> &Grammar {
+        self.dict.grammar()
+    }
+}
+
+impl<T> MorphemeList<T>
+where
+    T: Deref + Clone,
+    <T as Deref>::Target: DictionaryAccess,
+{
+    pub fn split(&self, mode: Mode, index: usize) -> SudachiResult<MorphemeList<T>> {
```

Comment:
As a general question, does it make sense to split only a single morpheme inside a path? (Except maybe tests)

Maybe the current Python API requires this, but we can provide better way to do things and deprecate bad ways later.
Usually we want to split the whole path at the same time, I believe.